tangle_develop captures wrong PID; quality gate aborts at 0% before providers finish

## Plugin Version

`9.29.2`

## Claude Code Version

`2.1.119`

## What happened?

Every `/octo:develop` (and any workflow that calls `tangle_develop`) aborts at the validation gate with `Quality gate FAILED with 0%`. The orchestrator reports `Progress: N/N subtasks complete` within ~1 second of dispatch — far too fast for any real provider call. Provider processes (codex/gemini) are still running 10+ minutes later (the heartbeat monitor keeps logging), but their result files are never read because `validate_tangle_results` fired immediately after the wait loop returned.

The decomposition step works correctly; the bug is specifically in the parallel-spawn wait loop.

Reduced reproduction (verified locally):

```
$ bash ~/.claude/plugins/cache/nyldn-plugins/octo/9.29.2/scripts/orchestrate.sh -v develop \
    'Write a one-line shell command that prints "hello".'
...
[INFO] Step 1: Task decomposition...     ← takes ~3 min, OK
[INFO] Step 2: Parallel execution...
[INFO] Spawned 1 development threads
Progress: 0/1 subtasks complete
[INFO] Spawning codex agent (task: tangle-1777317122-0, role: implementer)
[INFO] Agent spawned with PID: 35373
Progress: 1/1 subtasks complete           ← same second as the spawn
[INFO] Step 3: Validation gate...         ← fires before codex even starts thinking
[ERROR] Quality gate FAILED with 0%. Aborting workflow.

# 10 minutes later:
[DEBUG] Heartbeat for PID 35373 (still alive)   ← codex is still running
```

## What did you expect?

The wait loop should observe the actual provider PID and wait for it to exit before validation runs. After the fix described below: 1/1 subtasks complete arrives only after codex genuinely finishes; the validation gate sees the real result file; quality gate passes.

## Steps to reproduce

1. Have at least one external provider configured (codex or gemini OAuth/API key).
2. Run any `/octo:develop "<prompt>"` (a trivial prompt is enough).
3. Watch `Progress: N/N subtasks complete` arrive within ~1 second of `Spawning ... agent`.
4. `Quality gate FAILED with 0%` follows immediately.
5. `ps -ef | grep codex` (or gemini) — provider processes are still alive long after the workflow aborted.
6. `find ~/.claude/plugins/data/.../results -mmin -1` — no result files exist at the time validation ran; the providers eventually write them, but nothing reads them.

## Which providers are configured?

- Codex CLI
- Gemini CLI
- Perplexity API

(Bug reproduces with any of them; the dispatch path is shared.)

## Operating System

macOS (15.4 Sequoia / Apple silicon).

## Relevant logs

```shell
# scripts/lib/workflows.sh ~ line 793  (tangle_develop)

if [[ "$TMUX_MODE" == "true" ]]; then
    # Use async+tmux spawning  ← this branch is correct: pid=$(spawn_agent_async ...)
    local pid
    pid=$(spawn_agent_async "$agent" "$subtask" "$task_id" "$role" "tangle" "$pane_title")
    pids+=("$pid")
else
    # Standard spawning  ← BUG IS HERE
    spawn_agent "$agent" "$subtask" "$task_id" "$role" "tangle" &
    pids+=($!)
fi
```

`spawn_agent` (in `scripts/lib/spawn.sh:807`) launches the actual provider via `( ... ) &`, captures its PID, prints it to stdout via `echo "$pid"`, then returns. When `tangle_develop` runs `spawn_agent ... &`, the captured `$!` is the PID of the **outer wrapper subshell** that runs the spawn_agent function, not the inner provider PID. The wrapper exits within milliseconds (after launching the inner subshell and printing the PID), so the wait loop's `kill -0 $pid` instantly returns false and the subtask is marked "complete."

The TMUX branch already does the right thing because it uses command substitution to capture the PID that `spawn_agent_async` echoes.

---

## Proposed fix

A naive port — `pids+=($(spawn_agent ...))` — fixes correctness but breaks parallelism: bash command substitution waits for stdout EOF, and `spawn_agent`'s inner `( ... ) &` background subshell inherits stdout and keeps it open, so `$()` blocks until the actual provider exits. Verified locally: subtask N+1 only spawns when subtask N's provider exits.

The clean fix is a temp-file PID handoff: run `spawn_agent` in background with its stdout redirected to a tempfile, poll the tempfile for the PID, then continue the loop. The inner subshell still holds the fd open but we don't depend on EOF — we only need the PID for `kill -0` polling in the wait loop.

```diff
--- a/scripts/lib/workflows.sh
+++ b/scripts/lib/workflows.sh
@@ -793,8 +793,21 @@ tangle_develop() {
         else
             # Standard spawning
-            spawn_agent "$agent" "$subtask" "$task_id" "$role" "tangle" &
-            pids+=($!)
+            # Capture the inner provider PID (printed to stdout by spawn_agent
+            # at lib/spawn.sh:827) instead of $! of the outer wrapper.
+            local _pidfile
+            _pidfile=$(mktemp -t octo-tangle-pid.XXXXXX)
+            spawn_agent "$agent" "$subtask" "$task_id" "$role" "tangle" >"$_pidfile" 2>/dev/null &
+            local _wait=0
+            while [[ ! -s "$_pidfile" ]] && [[ $_wait -lt 100 ]]; do
+                sleep 0.1
+                ((_wait++)) || true
+            done
+            local _pid
+            _pid=$(grep -E '^[0-9]+$' "$_pidfile" 2>/dev/null | tail -1)
+            rm -f "$_pidfile"
+            if [[ -n "$_pid" ]]; then
+                pids+=("$_pid")
+            else
+                log WARN "spawn_agent produced no PID for $task_id within 10s — subtask may be lost"
+            fi
         fi
```

Verified locally: with this patch applied, parallel subtasks run concurrently, the wait loop blocks on the real provider PIDs, validation runs after all providers complete, and the quality gate sees actual result files. (My run still failed downstream because of Gemini rate-limit retries hammering the small subtask budget — that's a separate orchestration concern, but unrelated to this PID race.)

The patch lives at `~/.claude/patches/octo/0001-tangle-develop-capture-inner-provider-pid.patch` if helpful for a PR.

--- a/scripts/lib/agent-sync.sh
+++ b/scripts/lib/agent-sync.sh
@@ -188,7 +188,15 @@ ${provider_ctx}"
 
     # SECURITY: Use array-based execution to prevent word-splitting vulnerabilities
     local -a cmd_array
-    read -ra cmd_array <<< "$cmd"
+    local -a inner_cmd_array
+    build_provider_env "$agent_type"
+    read -ra inner_cmd_array <<< "$cmd"
+    if [[ ${#PROVIDER_ENV_ARRAY[@]} -gt 0 ]]; then
+        cmd_array=("${PROVIDER_ENV_ARRAY[@]}" "${inner_cmd_array[@]}")
+        log "DEBUG" "Credential isolation active for $agent_type"
+    else
+        cmd_array=("${inner_cmd_array[@]}")
+    fi
 
     # Capture output and exit code separately
     local output

--- a/scripts/lib/agent-utils.sh
+++ b/scripts/lib/agent-utils.sh
@@ -537,8 +537,8 @@ $prompt"
             ((subtask_num++)) || true
         else
             # Legacy bash subprocess
-            spawn_agent "$agent" "$prompt" "$retry_task_id" "$role" "tangle" &
-            local pid=$!
+            local pid
+            pid=$(spawn_agent_capture_pid "$agent" "$prompt" "$retry_task_id" "$role" "tangle")
             pids="$pids $pid"
             ((subtask_num++)) || true
             ((pid_count++)) || true
@@ -700,4 +700,3 @@ resume_agent() {
 }
 
 # [EXTRACTED to lib/spawn.sh]
-

--- a/scripts/lib/provider-routing.sh
+++ b/scripts/lib/provider-routing.sh
@@ -23,43 +23,45 @@
 
 # ═══════════════════════════════════════════════════════════════════════════════
 # SECURITY: Environment isolation for external CLI providers (v8.7.0)
-# Returns env prefix that limits environment variables to essentials only
+# Populates PROVIDER_ENV_ARRAY with argv tokens that limit environment
+# variables to essentials only. This stays safe when PATH contains spaces.
 # ═══════════════════════════════════════════════════════════════════════════════
 build_provider_env() {
     local provider="$1"
+    PROVIDER_ENV_ARRAY=()
 
     if [[ "${OCTOPUS_SECURITY_V870:-true}" != "true" ]]; then
         return 0
     fi
 
-    # v9.15.1: Skip env -i isolation on Windows — PATH contains spaces
-    # (C:\Program Files\...) which break read -ra word splitting in spawn.sh
-    case "${OCTOPUS_PLATFORM:-$(uname)}" in
-        MINGW*|MSYS*|CYGWIN*|Windows*) return 0 ;;
-    esac
-
     # v9.23: Propagate W3C trace headers into isolated env when present so
     # external CLIs (codex/gemini/perplexity) participate in distributed traces.
     # SUPPORTS_TRACEPARENT was detected in v2.1.98+ (Bash subprocesses) and
     # v2.1.110+ added the same for SDK/headless sessions.
-    local _trace_prefix=""
+    local -a _trace_env=()
     if [[ -n "${TRACEPARENT:-}" ]]; then
-        _trace_prefix+=" TRACEPARENT=${TRACEPARENT}"
+        _trace_env+=("TRACEPARENT=${TRACEPARENT}")
     fi
     if [[ -n "${TRACESTATE:-}" ]]; then
-        _trace_prefix+=" TRACESTATE=${TRACESTATE}"
+        _trace_env+=("TRACESTATE=${TRACESTATE}")
     fi
 
     # v9.2.1: Try resolving env vars before building isolated env (Issue #177)
     case "$provider" in
         codex*)
             [[ -z "${OPENAI_API_KEY:-}" ]] && resolve_provider_env "OPENAI_API_KEY" 2>/dev/null
-            echo "env -i PATH=$PATH HOME=$HOME OPENAI_API_KEY=${OPENAI_API_KEY:-} TMPDIR=${TMPDIR:-/tmp}${_trace_prefix}"
+            PROVIDER_ENV_ARRAY=(env -i "PATH=$PATH" "HOME=$HOME" "OPENAI_API_KEY=${OPENAI_API_KEY:-}" "TMPDIR=${TMPDIR:-/tmp}")
+            if [[ ${#_trace_env[@]} -gt 0 ]]; then
+                PROVIDER_ENV_ARRAY+=("${_trace_env[@]}")
+            fi
             ;;
         gemini*)
             [[ -z "${GEMINI_API_KEY:-}" ]] && resolve_provider_env "GEMINI_API_KEY" 2>/dev/null
             [[ -z "${GOOGLE_API_KEY:-}" ]] && resolve_provider_env "GOOGLE_API_KEY" 2>/dev/null
-            echo "env -i PATH=$PATH HOME=$HOME GEMINI_API_KEY=${GEMINI_API_KEY:-} GOOGLE_API_KEY=${GOOGLE_API_KEY:-} NODE_NO_WARNINGS=1 TMPDIR=${TMPDIR:-/tmp}${_trace_prefix}"
+            PROVIDER_ENV_ARRAY=(env -i "PATH=$PATH" "HOME=$HOME" "GEMINI_API_KEY=${GEMINI_API_KEY:-}" "GOOGLE_API_KEY=${GOOGLE_API_KEY:-}" "NODE_NO_WARNINGS=1" "TMPDIR=${TMPDIR:-/tmp}")
+            if [[ ${#_trace_env[@]} -gt 0 ]]; then
+                PROVIDER_ENV_ARRAY+=("${_trace_env[@]}")
+            fi
             ;;
         perplexity*)
             # perplexity_execute is a shell function — env -i cannot exec it (#300)
@@ -539,4 +541,3 @@ is_api_based_provider() {
             ;;
     esac
 }
-

--- a/scripts/lib/spawn.sh
+++ b/scripts/lib/spawn.sh
@@ -463,13 +463,14 @@ ${heuristic_ctx}"
         # SECURITY: Use array-based execution to prevent word-splitting vulnerabilities
         # v8.32.0: Per-provider credential isolation — each agent only sees its own API key
         local -a cmd_array
-        local env_prefix
-        env_prefix=$(build_provider_env "$agent_type")
-        if [[ -n "$env_prefix" ]]; then
-            read -ra cmd_array <<< "$env_prefix $cmd"
+        local -a inner_cmd_array
+        build_provider_env "$agent_type"
+        read -ra inner_cmd_array <<< "$cmd"
+        if [[ ${#PROVIDER_ENV_ARRAY[@]} -gt 0 ]]; then
+            cmd_array=("${PROVIDER_ENV_ARRAY[@]}" "${inner_cmd_array[@]}")
             log "DEBUG" "Credential isolation active for $agent_type"
         else
-            read -ra cmd_array <<< "$cmd"
+            cmd_array=("${inner_cmd_array[@]}")
         fi
 
         # IMPROVED: Use temp files for reliable output capture (v7.13.2 - Issue #10)
@@ -824,5 +825,42 @@ ${heuristic_ctx}"
     fi
 
     log INFO "Agent spawned with PID: $pid"
+    echo "$pid"
+}
+
+# Launch spawn_agent in the background and return the inner provider PID that
+# spawn_agent prints, not the short-lived wrapper PID from `$!`.
+spawn_agent_capture_pid() {
+    local agent_type="$1"
+    local prompt="$2"
+    local task_id="${3:-$(date +%s)}"
+    local role="${4:-}"
+    local phase="${5:-}"
+    local use_fork="${6:-false}"
+
+    local pid_file
+    pid_file=$(mktemp "${TMPDIR:-/tmp}/octo-spawn-pid.XXXXXX") || return 1
+
+    spawn_agent "$agent_type" "$prompt" "$task_id" "$role" "$phase" "$use_fork" >"$pid_file" 2>&1 &
+    local wrapper_pid=$!
+
+    local pid=""
+    local attempts=0
+    while [[ $attempts -lt 100 ]]; do
+        pid=$(awk '/^[0-9]+$/ { value=$1 } END { print value }' "$pid_file" 2>/dev/null)
+        [[ -n "$pid" ]] && break
+        if ! kill -0 "$wrapper_pid" 2>/dev/null && [[ -s "$pid_file" ]]; then
+            break
+        fi
+        sleep 0.1
+        ((attempts++)) || true
+    done
+
+    if [[ -z "$pid" ]]; then
+        log "WARN" "spawn_agent produced no provider PID for $task_id within 10s; tracking wrapper PID $wrapper_pid" >&2
+        pid="$wrapper_pid"
+    fi
+
+    rm -f "$pid_file"
     echo "$pid"
 }

--- a/scripts/lib/workflows.sh
+++ b/scripts/lib/workflows.sh
@@ -121,12 +121,13 @@ IMPORTANT: If you find yourself searching or grepping more than 3 times in a row
 
     # Build command array with credential isolation
     local -a cmd_array
-    local env_prefix
-    env_prefix=$(build_provider_env "$agent_type")
-    if [[ -n "$env_prefix" ]]; then
-        read -ra cmd_array <<< "$env_prefix $cmd"
+    local -a inner_cmd_array
+    build_provider_env "$agent_type"
+    read -ra inner_cmd_array <<< "$cmd"
+    if [[ ${#PROVIDER_ENV_ARRAY[@]} -gt 0 ]]; then
+        cmd_array=("${PROVIDER_ENV_ARRAY[@]}" "${inner_cmd_array[@]}")
     else
-        read -ra cmd_array <<< "$cmd"
+        cmd_array=("${inner_cmd_array[@]}")
     fi
 
     local temp_output="${RESULTS_DIR}/.tmp-${task_id}.out"
@@ -443,8 +444,9 @@ ${_blind_spot_checklist}"
             pids+=("$pid")
         else
             # Standard spawning
-            spawn_agent "$agent" "$perspective" "$task_id" "researcher" "probe" &
-            pids+=($!)
+            local pid
+            pid=$(spawn_agent_capture_pid "$agent" "$perspective" "$task_id" "researcher" "probe")
+            pids+=("$pid")
         fi
         sleep 0.1
     done
@@ -792,8 +794,9 @@ Output as numbered list with [CODING] or [REASONING] prefix for each subtask."
             pids+=("$pid")
         else
             # Standard spawning
-            spawn_agent "$agent" "$subtask" "$task_id" "$role" "tangle" &
-            pids+=($!)
+            local pid
+            pid=$(spawn_agent_capture_pid "$agent" "$subtask" "$task_id" "$role" "tangle")
+            pids+=("$pid")
         fi
         ((subtask_num++)) || true
     done <<< "$subtasks"

--- a/scripts/lib/yaml-workflow.sh
+++ b/scripts/lib/yaml-workflow.sh
@@ -341,10 +341,7 @@ $previous_output"
                 fi
                 sleep 2
             done
-            # Wait for remaining pids to avoid zombies
-            for pid in "${pids[@]}"; do
-                wait "$pid" 2>/dev/null || true
-            done
+            _yaml_wait_for_pids "$max_wait" "${pids[@]}"
         else
             _yaml_wait_for_pids "${TIMEOUT:-600}" "${pids[@]}"
         fi

--- a/scripts/lib/yaml-workflow.sh
+++ b/scripts/lib/yaml-workflow.sh
@@ -168,6 +168,29 @@ resolve_prompt_template() {
     echo "$resolved"
 }
 
+_yaml_wait_for_pids() {
+    local max_wait="${1:-${TIMEOUT:-600}}"
+    shift || true
+    local pids=("$@")
+    local wait_start=$SECONDS
+
+    while [[ ${#pids[@]} -gt 0 && $(( SECONDS - wait_start )) -lt $max_wait ]]; do
+        local all_done=true
+        local pid
+        for pid in "${pids[@]}"; do
+            [[ -z "$pid" ]] && continue
+            if kill -0 "$pid" 2>/dev/null; then
+                all_done=false
+                break
+            fi
+        done
+        [[ "$all_done" == "true" ]] && return 0
+        sleep 2
+    done
+
+    return 0
+}
+
 # Execute a single workflow phase from YAML definition
 # Spawns agents as defined, respects parallel/sequential flags, evaluates quality gates
 execute_workflow_phase() {
@@ -276,15 +299,14 @@ $previous_output"
         esac
 
         if [[ "$is_parallel" == "true" ]]; then
-            spawn_agent "$agent_type" "$agent_prompt" "$task_id" "$role" "$phase_name" &
-            pids+=($!)
+            local pid
+            pid=$(spawn_agent_capture_pid "$agent_type" "$agent_prompt" "$task_id" "$role" "$phase_name")
+            pids+=("$pid")
         else
             # Sequential agent - wait for parallel agents first
             if [[ ${#pids[@]} -gt 0 ]]; then
                 log "DEBUG" "Waiting for ${#pids[@]} parallel agents before sequential agent"
-                for pid in "${pids[@]}"; do
-                    wait "$pid" 2>/dev/null || true
-                done
+                _yaml_wait_for_pids "${TIMEOUT:-600}" "${pids[@]}"
                 pids=()
             fi
             spawn_agent "$agent_type" "$agent_prompt" "$task_id" "$role" "$phase_name"
@@ -324,9 +346,7 @@ $previous_output"
                 wait "$pid" 2>/dev/null || true
             done
         else
-            for pid in "${pids[@]}"; do
-                wait "$pid" 2>/dev/null || true
-            done
+            _yaml_wait_for_pids "${TIMEOUT:-600}" "${pids[@]}"
         fi
     fi
 

--- a/tests/smoke/test-fleet-dispatch-guard.sh
+++ b/tests/smoke/test-fleet-dispatch-guard.sh
@@ -100,7 +100,21 @@ test_provider_pid_capture() {
     done
 
     local wrapper_tracking
-    wrapper_tracking=$(grep -Rn 'pids.*\$!' "$PROJECT_ROOT/scripts/lib/workflows.sh" "$PROJECT_ROOT/scripts/lib/yaml-workflow.sh" "$PROJECT_ROOT/scripts/lib/agent-utils.sh" 2>/dev/null || true)
+    wrapper_tracking=$(
+        {
+            grep -RnE 'pids[^=]*[+]?=.*\$!' \
+                "$PROJECT_ROOT/scripts/lib/workflows.sh" \
+                "$PROJECT_ROOT/scripts/lib/yaml-workflow.sh" \
+                "$PROJECT_ROOT/scripts/lib/agent-utils.sh" 2>/dev/null || true
+            awk '
+                /pid=\$!/ { pending=NR; line=$0; next }
+                pending && NR <= pending + 5 && /pids/ { print FILENAME ":" pending ":" line " ... " $0; pending=0 }
+                pending && NR > pending + 5 { pending=0 }
+            ' "$PROJECT_ROOT/scripts/lib/workflows.sh" \
+              "$PROJECT_ROOT/scripts/lib/yaml-workflow.sh" \
+              "$PROJECT_ROOT/scripts/lib/agent-utils.sh" 2>/dev/null || true
+        } | sed '/^$/d'
+    )
     if [[ -n "$wrapper_tracking" ]]; then
         echo "  Found wrapper PID tracking:"
         echo "$wrapper_tracking"

--- a/tests/smoke/test-fleet-dispatch-guard.sh
+++ b/tests/smoke/test-fleet-dispatch-guard.sh
@@ -83,8 +83,36 @@ EOF
     fi
 }
 
+test_provider_pid_capture() {
+    test_case "Parallel workflow waits track provider PIDs, not wrapper PIDs"
+
+    local failed=0
+    if ! grep -q '^spawn_agent_capture_pid()' "$PROJECT_ROOT/scripts/lib/spawn.sh"; then
+        echo "  MISSING spawn_agent_capture_pid helper in scripts/lib/spawn.sh"
+        failed=1
+    fi
+
+    for rel in scripts/lib/workflows.sh scripts/lib/yaml-workflow.sh scripts/lib/agent-utils.sh; do
+        if ! grep -q 'spawn_agent_capture_pid' "$PROJECT_ROOT/$rel"; then
+            echo "  MISSING provider PID capture in: $rel"
+            failed=1
+        fi
+    done
+
+    local wrapper_tracking
+    wrapper_tracking=$(grep -Rn 'pids.*\$!' "$PROJECT_ROOT/scripts/lib/workflows.sh" "$PROJECT_ROOT/scripts/lib/yaml-workflow.sh" "$PROJECT_ROOT/scripts/lib/agent-utils.sh" 2>/dev/null || true)
+    if [[ -n "$wrapper_tracking" ]]; then
+        echo "  Found wrapper PID tracking:"
+        echo "$wrapper_tracking"
+        failed=1
+    fi
+
+    [[ $failed -eq 0 ]] && test_pass || test_fail "Parallel workflow code still risks tracking short-lived spawn_agent wrapper PIDs"
+}
+
 test_fleet_guard_present
 test_no_bare_force_legacy_export
 test_hooks_json_matchers
+test_provider_pid_capture
 
 test_summary

--- a/tests/test-credential-isolation.sh
+++ b/tests/test-credential-isolation.sh
@@ -43,7 +43,7 @@ else
 fi
 
 # 1.2 Codex scoping — only OPENAI_API_KEY
-CODEX_ENV=$(grep -A5 'codex\*)' "$ALL_SRC" | grep 'env -i' | head -1)
+CODEX_ENV=$(grep -A5 'codex\*)' "$ALL_SRC" | grep 'PROVIDER_ENV_ARRAY=.*env -i' | head -1)
 if echo "$CODEX_ENV" | grep -q 'OPENAI_API_KEY'; then
   pass "Codex env includes OPENAI_API_KEY"
 else
@@ -57,7 +57,7 @@ else
 fi
 
 # 1.3 Gemini scoping — only GEMINI_API_KEY + GOOGLE_API_KEY
-GEMINI_ENV=$(grep -A5 'gemini\*)' "$ALL_SRC" | grep 'env -i' | head -1)
+GEMINI_ENV=$(grep -A5 'gemini\*)' "$ALL_SRC" | grep 'PROVIDER_ENV_ARRAY=.*env -i' | head -1)
 if echo "$GEMINI_ENV" | grep -q 'GEMINI_API_KEY'; then
   pass "Gemini env includes GEMINI_API_KEY"
 else
@@ -74,6 +74,7 @@ fi
 # perplexity_execute is a bash function dispatched by get_agent_command();
 # env -i cannot exec shell functions, so build_provider_env returns empty.
 PERP_CASE=$(grep -A50 'build_provider_env()' "$ALL_SRC" | grep -A5 'perplexity\*)' | head -6)
+PERP_ENV=$(echo "$PERP_CASE" | grep 'env -i' | head -1 || true)
 if echo "$PERP_CASE" | grep -q 'resolve_provider_env.*PERPLEXITY_API_KEY'; then
   pass "Perplexity resolves PERPLEXITY_API_KEY before dispatch"
 else
@@ -84,6 +85,18 @@ if echo "$PERP_CASE" | grep -q 'return 0'; then
   pass "Perplexity correctly returns empty env prefix (shell function)"
 else
   fail "Perplexity should return 0 (no env -i for shell function provider)"
+fi
+
+if grep -q 'PROVIDER_ENV_ARRAY=()' "$ALL_SRC" && grep -q 'PROVIDER_ENV_ARRAY\[@\]' "$ALL_SRC"; then
+  pass "Provider env uses argv array tokens"
+else
+  fail "Provider env array token handling missing"
+fi
+
+if grep -A20 'build_provider_env()' "$ALL_SRC" | grep -q 'MINGW.*return 0\|MSYS.*return 0\|Windows.*return 0'; then
+  fail "Windows still disables env isolation instead of preserving PATH spaces with arrays"
+else
+  pass "Windows PATH spaces do not disable env isolation"
 fi
 
 # ─────────────────────────────────────────────────────────────────────

